particle[] particles
uint16 weightSum